core/modules/css/DbInterfaceMySql.cc

```diff
@@ -0,0 +1,126 @@
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the LSST License Statement and
+ * the GNU General Public License along with this program.  If not,
+ * see <http://www.lsstcorp.org/LegalNotices/>.
+ */
+
+// Class header
+#include "css/DbInterfaceMySql.h"
+
+// System headers
+#include <fstream>
+#include <iomanip>
+#include <iostream>
+#include <sstream>
+#include <stdexcept>
+
+// Third-party headers
+#include "mysql/mysqld_error.h"
+
+// LSST headers
+#include "lsst/log/Log.h"
+
+// Qserv headers
+#include "css/CssError.h"
+#include "css/DbInterfaceMySql.h"
+#include "sql/SqlResults.h"
+#include "sql/SqlTransactionStd.h"
+
+using namespace std;
+
+namespace {
+LOG_LOGGER _log = LOG_GET("lsst.qserv.css.DbInterfaceMySql");
+
+} // namespace
+
+namespace lsst {
+namespace qserv {
+namespace css {
+
+
+class KvTransaction : public sql::SqlTransactionStd
+{
+public:
+    // Constructors
+    KvTransaction(sql::SqlConnection& conn)
+        : sql::SqlTransactionStd(conn) {}
+
+    // Destructor
+    ~KvTransaction() override = default;
+
+    /// Override to throw an appropriate exception.
+    void throwException(util::Issue::Context const& ctx, std::string const& msg) override {
+        throw CssError(ctx, msg + " mysql(" +std::to_string(errObj.errNo()) + " " + errObj.errMsg() + ")");
+    }
+};
+
+
+DbInterfaceMySql::DbInterfaceMySql(mysql::MySqlConfig const& mysqlConf)
+: _conn(mysqlConf) {
+}
+
+
+std::set<int> DbInterfaceMySql::getEmptyChunks(std::string const& dbName) {
+    std::string funcName("getEmptyChunksFromDb");
+    std::lock_guard<std::mutex> sync(_dbMutex);
+
+    KvTransaction trans(_conn);
+
+    /// This must match the name of the table generated by
+    /// dataloader.py DataLoader _makeEmptyChunksTable()
+    std::string emptyChunkTbl = dbName + "_EmptyChunks";
+
+    // run query
+    sql::SqlErrorObject errObj;
+    sql::SqlResults results;
+    std::string const query = "SELECT * FROM " + emptyChunkTbl;
+    LOGS(_log, LOG_LVL_DEBUG, "Executing query: " << query);
+    if (not _conn.runQuery(query, results, errObj)) {
+        LOGS(_log, LOG_LVL_ERROR, funcName << " SQL query failed: " << query);
+        throw CssError(ERR_LOC, errObj);
+    }
+
+    // get results of the query
+    std::vector<std::string> emptyChunks;
+    if (not results.extractFirstColumn(emptyChunks, errObj)) {
+        LOGS(_log, LOG_LVL_ERROR, funcName << "Failed to extract empty chunks from query result");
+        throw CssError(ERR_LOC, errObj);
+    }
+
+    trans.commit();
+
+    // Create the set
+    std::set<int> emptyChunkSet;
+    for (auto j:emptyChunks) {
+        try {
+            emptyChunkSet.insert(std::stoi(j));
+        } catch (std::invalid_argument const& e) {
```

Comment:
Strictly speaking, you may also need to intercept the `out-of-band` exceptions when a value is not falling into the allowed range for type `int`. Then you would need to catch `std::out_of_range` as well.